main.js

```diff
@@ -0,0 +1,270 @@
+
+          // replace with new row/array to not modify the previous one.
+          return stateArr.board[index] = newRow;
+
+        } //If statement
+      }) // End forEach array
+    }
+
+    return stateArr
+  }
+  play(player, data) {
+    // invoke next method to check and raplce the users choice.
+    this.replaceChar(player, data)
+    // counter to check when the game has reached all nine turns
+    data.counter += 1;
+    // start checking for patterns to see if player wins.
+    if (data.counter >= 5) {
+      let playerWon = this.checkWinner(player);
+      // If player wins change its attributes to True.
+      if (playerWon) {
+        this.showBoard(data);
+        data.playerWins = true;
+        player.winner = true;
+      } else if(data.counter === 9) {
+        this.showBoard(data);
+        let playAgain = prompt('Tie, do you want to play again?, y or Y = yes, n or N to stop the game.').toLowerCase();
+        if (playAgain === 'y') {
+          main();
+        } else {
+          throw new Error("Game Over, thanks for playing");
+        }
+      }
+    }
+  } // End play
+
+  // start method
+  start(data) {
+    while (data.playerWins !== true) {
+      // show the (numbers)colums and rows of the game.
+      this.showBoard(data);
+      // get each player to play once, alternating after the other.
+      // as long the player has not won.
+      this.play(this.player1, data);
+      // this could be way better, will refactor after submitting
+      if (this.player1.winner === true) {
+        alert(`Congratulations ${this.player1.name} you won!`);
+        let playAgain = prompt('Do you want to play again?, y or Y = yes, n or N to stop the game.').toLowerCase();
+        if (playAgain === 'y') {
+          main();
+        } else {
+          throw new Error("Game Over, thanks for playing");
+        }
+      } else {
+        // show the (numbers)colums and rows of the game.
+        this.showBoard(data);
+        // get each player to play once, alternating after the other.
+        // as long the player has not won.
+        this.play(this.player2, data);
+        if (this.player2.winner === true) {
+          alert(`Congratulations ${this.player2.name} you won!`);
+          let playAgain = prompt('Do you want to play again?, y or Y = yes, n or N to stop the game.').toLowerCase();
+          if (playAgain === 'y') {
+            main();
+          } else {
+            throw new Error("Game Over, thanks for playing");
+          }
+        }
+      }
+    }
+
+  }
+
+
+  showBoard(data) {
+    let joinRows = data.board.map(function(row, index) {
+      return row.join('')
+    }) // End map array
+    // console.log(joinRows);
+
+    let formattedArray = joinRows.join('\n')
+    console.log(formattedArray);
+    // return formattedArray
+  }
+
+  // Algorithm borrowed from https://gist.github.com/axelpale/3118596.
+  // I modified it to work with the patterns of the TIC-TAC-TOE game.
+  k_combinations(set, k) {
```

Comment:
Cool, but what's `k`, `j`, `combs`, `head` and `tailcombs`? Perhaps the naming should be changed to give context here.